test/regress/regress-2034.txt

```diff
@@ -0,0 +1,17 @@
+;;; TOOL: run-spec-wasm2c
+;;; ARGS: --debug-names --enable-exceptions
+(module
+ (func $x)
+ (export "func" (func $x))
+ (global $x i32 (i32.const 14))
+ (table $x (export "tab") 0 funcref)
+ (elem $x)
+ (memory $x (export "mem") 0)
+ (data $x)
+ (tag $x (export "tag"))
```

Comment:
Do we really not have any spec tests that include elements of different types with the same name?   Its seems odd that we would need to write a wasm2c specific test for this behaviour.  Or is it because we don't run the spec tests with --debug-names enabled normally?

It does seem odd to be that this should even be allowed in the spec.  Wouldn't it be simple to force all names to be globally unique?

Reply:
Sadly it does looks like identifiers are only unique per-index-space: https://webassembly.github.io/spec/core/text/values.html#text-id

Reply:
Sure, do you want that in the test or just here in GitHub?

wasm2c was already good about keeping the C identifiers unique (`DefineName` takes care of that), so the output ends up looking like this:

```c
typedef struct Z_regress_2034_0_wasm_instance_t {
  u32 w2c_foo;
  wasm_rt_memory_t w2c_foo_0;
  wasm_rt_funcref_table_t w2c_foo_1;
} Z_regress_2034_0_wasm_instance_t;
```

```
static u32 w2c_foo_4;

static void init_tags(void) {
  w2c_foo_4 = wasm_rt_register_tag(0);
}

static void w2c_foo_5(Z_regress_2034_0_wasm_instance_t*);

static void init_globals(Z_regress_2034_0_wasm_instance_t* instance) {
  instance->w2c_foo = 14u;
}


static void init_memories(Z_regress_2034_0_wasm_instance_t* instance) {
  wasm_rt_allocate_memory(&instance->w2c_foo_0, 0, 65536);
}

static void init_data_instances(Z_regress_2034_0_wasm_instance_t *instance) {
}

static void init_tables(Z_regress_2034_0_wasm_instance_t* instance) {
  wasm_rt_allocate_funcref_table(&instance->w2c_foo_1, 0, 4294967295);
}

static void init_elem_instances(Z_regress_2034_0_wasm_instance_t *instance) {
}

static void w2c_foo_5(Z_regress_2034_0_wasm_instance_t* instance) {
  FUNC_PROLOGUE;
  FUNC_EPILOGUE;
}

/* export: 'func' */
void Z_regress_2034_0_wasmZ_func(Z_regress_2034_0_wasm_instance_t* instance) {
  return w2c_foo_5(instance);
}

/* export: 'tab' */
wasm_rt_funcref_table_t* Z_regress_2034_0_wasmZ_tab(Z_regress_2034_0_wasm_instance_t* instance) {
  return &instance->w2c_foo_1;
}

/* export: 'mem' */
wasm_rt_memory_t* Z_regress_2034_0_wasmZ_mem(Z_regress_2034_0_wasm_instance_t* i
nstance) {
  return &instance->w2c_foo_0;
}

/* export: 'tag' */
const u32 *Z_regress_2034_0_wasmZ_tag = &w2c_foo_4;
```

Reply:
Or if you prefer the diff:

```diff
@@ -585,7 +585,7 @@
 static u32 w2c_foo_4;
 
 static void init_tags(void) {
-  w2c_foo = wasm_rt_register_tag(0);
+  w2c_foo_4 = wasm_rt_register_tag(0);
 }
 
 static void w2c_foo_5(Z_regress_2034_0_wasm_instance_t*);
@@ -596,41 +596,41 @@
 
 
 static void init_memories(Z_regress_2034_0_wasm_instance_t* instance) {
-  wasm_rt_allocate_memory(&instance->w2c_foo, 0, 65536);
+  wasm_rt_allocate_memory(&instance->w2c_foo_0, 0, 65536);
 }
 
 static void init_data_instances(Z_regress_2034_0_wasm_instance_t *instance) {
 }
 
 static void init_tables(Z_regress_2034_0_wasm_instance_t* instance) {
-  wasm_rt_allocate_funcref_table(&instance->w2c_foo, 0, 4294967295);
+  wasm_rt_allocate_funcref_table(&instance->w2c_foo_1, 0, 4294967295);
 }
 
 static void init_elem_instances(Z_regress_2034_0_wasm_instance_t *instance) {
 }
 
-static void w2c_foo(Z_regress_2034_0_wasm_instance_t* instance) {
+static void w2c_foo_5(Z_regress_2034_0_wasm_instance_t* instance) {
   FUNC_PROLOGUE;
   FUNC_EPILOGUE;
 }
 
 /* export: 'func' */
 void Z_regress_2034_0_wasmZ_func(Z_regress_2034_0_wasm_instance_t* instance) {
-  return w2c_foo(instance);
+  return w2c_foo_5(instance);
 }
 
 /* export: 'tab' */
 wasm_rt_funcref_table_t* Z_regress_2034_0_wasmZ_tab(Z_regress_2034_0_wasm_instance_t* instance) {
-  return &instance->w2c_foo;
+  return &instance->w2c_foo_1;
 }
 
 /* export: 'mem' */
 wasm_rt_memory_t* Z_regress_2034_0_wasmZ_mem(Z_regress_2034_0_wasm_instance_t* instance) {
-  return &instance->w2c_foo;
+  return &instance->w2c_foo_0;
 }
 
 /* export: 'tag' */
-const u32 *Z_regress_2034_0_wasmZ_tag = &w2c_foo;
+const u32 *Z_regress_2034_0_wasmZ_tag = &w2c_foo_4;
 
 void Z_regress_2034_0_wasm_init_module(void) {
   assert(wasm_rt_is_initialized());
@@ -650,6 +650,6 @@
 }
 
 void Z_regress_2034_0_wasm_free(Z_regress_2034_0_wasm_instance_t* instance) {
-  wasm_rt_free_funcref_table(&instance->w2c_foo);
-  wasm_rt_free_memory(&instance->w2c_foo);
+  wasm_rt_free_funcref_table(&instance->w2c_foo_1);
+  wasm_rt_free_memory(&instance->w2c_foo_0);
 }
```
```

Reply:
> Do we really not have any spec tests that include elements of different types with the same name? Its seems odd that we would need to write a wasm2c specific test for this behaviour. Or is it because we don't run the spec tests with --debug-names enabled normally?

My current understanding is that:
- the spec tests don't always use symbolic identifiers at all (maybe in half of the tests?)
- when they do, they never reuse the same identifier to refer to different elements within a module (the wasm2c tests pass 100% even with `--debug-names` turned on)
- and even if they did, we wouldn't have caught it because we don't run the wasm2c spec tests with `--debug-names` enabled

If you want to turn on `--debug-names` be default for wasm2c testing, I wouldn't have any objection, although the real way I hit this bug was from names coming out of the NameGenerator (since it generates `$e0` both for tag #0 and for elem segment #0) so I think we'd lose some test coverage either way.